src/main/java/com/monitorjbl/plugins/AsyncProcessor.java

```diff
@@ -1,32 +1,48 @@
 package com.monitorjbl.plugins;
 
-import com.atlassian.bitbucket.concurrent.BucketProcessor;
+import com.atlassian.bitbucket.util.Version;
+import com.atlassian.bitbucket.util.concurrent.ExecutorUtils;
+import com.atlassian.sal.api.ApplicationProperties;
+import com.atlassian.sal.api.lifecycle.LifecycleAware;
+import com.atlassian.util.concurrent.ThreadFactories;
+import org.slf4j.LoggerFactory;
 
-import javax.annotation.Nonnull;
-import java.io.Serializable;
-import java.util.List;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
-public class AsyncProcessor {
-  private static final String PREFIX = "com.monitorjbl.plugins:pr-harmony:";
-  private final ExecutorService concurrencyService;
+public class AsyncProcessor implements LifecycleAware {
 
-  public AsyncProcessor(ExecutorService concurrencyService) {
-    this.concurrencyService = concurrencyService;
-  }
+  private static final Version VERSION_5_9 = new Version(5, 9);
 
-  @SuppressWarnings("unchecked")
-  public <T extends Serializable> void dispatch(Runnable taskRequest) {
-    concurrencyService.submit(taskRequest);
-  }
+  private final ExecutorService executorService;
+  private final boolean ownsExecutor;
 
-  public static abstract class TaskProcessor<T extends Serializable> implements BucketProcessor<T> {
-    @Override
-    public void process(@Nonnull String s, @Nonnull List<T> list) {
-      list.forEach(this::handleTask);
+  public AsyncProcessor(ApplicationProperties applicationProperties, ExecutorService executorService) {
+    ownsExecutor = new Version(applicationProperties.getVersion()).compareTo(VERSION_5_9) < 0;
+    if (ownsExecutor) {
+      //For Bitbucket Server versions before 5.9, use our own ExecutorService to avoid starving out hook callbacks
+      this.executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(),
+          ThreadFactories.namedThreadFactory("pr-harmony", ThreadFactories.Type.DAEMON));
+    } else {
+      //Bitbucket Server 5.9 includes a fix for BSERV-10652, which moves hook callback to their own threadpool, so
+      //for 5.9+ use the shared executor
+      this.executorService = executorService;
```

Comment:
For 5.9, this allows PR Harmony to continue using our shared `ExecutorService`. You don't have to do it this way (you could _always_ use your own), but I'd prefer (and appreciate!) it if you did. This way the background processing PR Harmony triggers "shares" the server with other background processing, rather than always running in addition to it. That should help manage overall server load.

Reply:
Will this not cause thread exhaustion when the server gets a lot of build updates? If the executor wasn't shared, it wouldn't directly interfere with that (although it certainly could interfere with CPU/memory consumption).

Reply:
This is a (difficult) balancing act.

Certainly, having a significant number of build status changes could result in PR Harmony creating a lot of tasks for the shared executor. In a system where there are many add-ons using the shared executor, those build status tasks may delay tasks created by other add-ons--just like a large backlog of tasks from other add-ons would be able to delay PR Harmony's build status processing.

For Bitbucket Server, though, that's actually _desirable_, as long as the tasks being delayed are not critical. That's the real "bug" in BSERV-10652; we were using the shared executor for _critical_ processing: Hook callbacks. Having a shared limit on background processing that can be delayed helps to ensure the accumulation of add-ons doesn't overwhelm the system.

(That's not to say that PR Harmony's build status processing is not _important_; clearly it is. But a delay of minutes is less critical than with, say, hosting operations.)

Now, it could certainly be contended that if PR Harmony just always used its own executor, it wouldn't starve critical processing, but it also wouldn't starve, or be starved by, other add-ons. From a purely PR Harmony standpoint, that's all good. But from _my_ standpoint, which I'll readily admit is biased, it's less good. The build status processing PR Harmony does is quite heavy, and produces non-trivial system load. When load becomes too high, it's usually not add-on vendors that see the support load, at least initially; it's Bitbucket Server's developers and supporters.

If all add-ons create and use their own executors, it's likely some (most? all?) of them will use non-configurable sizes (as I've done here, I'll note, for 5.8 and earlier). But even if they _are_ configurable, that still means a system administrator has to find out what all those configuration properties are and apply them--and they'd need to have some way of readily determining which add-on(s) executor(s) they need to reconfigure based on who's starving (or who's creating too much load). That's not always straightforward to figure out, especially for non-programmers.

By comparison, if add-ons use the shared executor, it's a single thing to size. That implies, if tasks are getting backed up, administrators can consider their current system load and opt to add more threads to the shared executor, which would then be "spread" across all the add-ons that use it and allow the backlog to be cleared more quickly. Additionally, we're planning to add monitoring for the queue size on the shared executor in a future release of Bitbucket Server, which will make it easier for administrators to track the size of the backlog on their server. That's another feature add-ons which use their own executors won't benefit from. We may add an SPI that they could use (not sure what the plan is there, at the moment), but they'd still have to manually build it out.